src/main/java/com/google/sps/data/AvailabilityTimeSlotGenerator.java

```diff
@@ -0,0 +1,158 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.TextStyle;
+import java.util.Locale;
+import java.time.format.DateTimeFormatter;
+
+// TODO: Add a test for the constructor.
+
+/** A generator of a collection of AvailabilityTimeSlot Objects. */
+public class AvailabilityTimeSlotGenerator {
+  private static final int numberOfSlotsPerDay = 48;
+  private static final int TIME_8AM = 8;
+  private static final int TIME_7PM = 19;
+
+  /**
+   * Constructs a list of a day's worth of AvailabilityTimeSlot objects.
+   *
+   * @param instant An instant on the day that time slots are generate for.
+   * @param timezoneOffset A String that represents the difference between UTC and the user's
+   *     current timezone. Example: A user in EST has a timezoneOffset of "-240" which means that
+   *     EST is 240 minutes behind UTC.
+   */
+  // TODO: Create a timeSlotsForWeek method.
+  public static List<AvailabilityTimeSlot> timeSlotsForDay(Instant instant, String timezoneOffset) {
+    ZonedDateTime day = generateDay(instant, timezoneOffset);
+    ZoneId zoneId = day.getZone();
+    String dayOfWeek = day.getDayOfWeek().getDisplayName(TextStyle.SHORT, Locale.US);
+    int year = day.getYear();
+    int month = day.getMonthValue();
+    int dayOfMonth = day.getDayOfMonth();
+
+    String date = generateDate(dayOfWeek, month, dayOfMonth);
+    List<String> times = generateTimes();
+    List<String> utcEncodings = generateUTCEncodings(year, month, dayOfMonth, zoneId);
+    List<Boolean> selectedStatuses = getSelectedStatuses();
+    return generateTimeSlots(utcEncodings, times, date, selectedStatuses);
+  }
+
+  // Uses an Instant and a timezoneOffset String to create a ZonedDateTime instance
+  // for the day specified by the Instant.
+  private static ZonedDateTime generateDay(Instant instant, String timezoneOffset) {
+    ZoneOffset timeZoneOffset = convertStringToOffset(timezoneOffset);
+    ZoneId zoneId = ZoneId.ofOffset("UTC", timeZoneOffset);
+    return instant.atZone(zoneId);
+  }
+
+  // This method takes the timezoneOffset String and converts it
+  // into a proper ZoneOffset instance.
+  private static ZoneOffset convertStringToOffset(String timezoneOffset) {
+    int offsetTotalMinutes = Integer.parseInt(timezoneOffset);
+    int offsetHours = offsetTotalMinutes / 60;
+    int offsetMinutes = offsetTotalMinutes % 60;
+    return ZoneOffset.ofHoursMinutes(offsetHours, offsetMinutes);
+  }
+
+  // Returns a readable date string such as "Tue 7/7".
+  private static String generateDate(String dayOfWeek, int month, int dayOfMonth) {
+    return dayOfWeek + " " + month + "/" + dayOfMonth;
+  }
+
+  // This method returns a list of UTC Strings for a single day, creating one String per Time
+  // Slot.
+  private static List<String> generateUTCEncodings(
```

Comment:
This function duplicates the logic in generateTimes. Can it be unified with that function?

Reply:
I think they're different enough that their purpose becomes clearer with separate functions.

Reply:
Discussed offline,

The real thing I should've gotten at here is that we're iterating a bunch of times where we really only need to iterate once.

Here's a sketch of the relevant code

```java
String formattedDate = formatDate(dayOfWeek, month, dayOfMonth);
ImmutableList.Builder<AvailabilityTimeSlot> slots = ImmutableList.builder();
for (Pair<Integer, Integer> hourAndMinute : candidateHoursAndMinutes) {
  int hour = hourAndMinute.getFirst();
  int minute = hourAndMinute.getSecond();
  LocalDateTime localTime = LocalDateTime.of(year, month, dayOfMonth, hour, minute);
  ZonedDateTime utcTime = ZonedDateTime.of(localTime, zoneId).withZoneSameInstant(ZoneOffset.UTC);
  String formattedUtc = utcTimeSlot.format(DateTimeFormatter.ISO_INSTANT);

  int standardHour = hour;
  if (hour > 12) {
    standardHour = hour - 12;
  }
  String formattedTime = String.format("%d:%02d %s", standardHour, minute, hour < 12 ? "AM" : "PM"));
  slots.add(
    AvailabilityTimeSlot.create(
        utcTime.format(DateTimeFormatter.ISO_INSTANT),
        formattedTime,
        formattedDate));
}
return slots.build();
```